### PR TITLE
✨ Introduce create-only bit in downsync API

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
## Summary
This PR introduces the API changes for a feature whereby a user can mark a workload object for propagation in "create-only" mode (#2336).

I plan to update the binding controller and generic transport controller in later PRs.

This feature can not be completed until a release of this repo is made, then changes are made in https://github.com/kubestellar/ocm-transport-plugin (https://github.com/kubestellar/ocm-transport-plugin/issues/34) using this new API, then a release of that repo, then this repo is updated to use that release.

## Related issue(s)

This addresses part of #2336.

### CODE REVIEW

File: pkg/transport/generic_transport_controller.go [hunk 1]
```diff
@@ -847,8 +852,19 @@ func (c *genericTransportController) computeDestToCustomizedObjects(objectsToPro
 	return destToCustomizedObjects, bindingErrors
 }
 
+func TODO_FIXME_ADD_CONSTANT_CREATEONLY(obj *unstructured.Unstructured) Wrapee {
```
Comment (vMaroon) [inline pkg/transport/generic_transport_controller.go:855]: FYI: The standard is to leave a comment that includes "TODO", which is followed by modern IDEs to list pending work.

File: pkg/transport/generic_transport_controller.go [hunk 2]
```diff
@@ -847,8 +852,19 @@ func (c *genericTransportController) computeDestToCustomizedObjects(objectsToPro
...
+	return Wrapee{obj, false}
+}
+
 func (c *genericTransportController) wrapBatch(batchToPropagate []*unstructured.Unstructured, binding *v1alpha1.Binding, numShard int, isSharded bool) (*unstructured.Unstructured, error) {
-	wrappedObject, err := convertObjectToUnstructured(c.transport.WrapObjects(batchToPropagate))
+	var wrapped runtime.Object
+	if t2, is := c.transport.(TransportWithCreateOnly); is {
```
Comment (vMaroon) [inline pkg/transport/generic_transport_controller.go:861]: Does this not beat the point of "generic" in `genericTransportController`? Shouldn't the difference in functionality be abstracted in the interface and implemented twice?

Due to the urgency (if i'm not mistaken) - it is fine to keep it as-is but add a TODO or document this in any way - that is assuming you agree.

File: pkg/transport/transport.go [hunk 3]
```diff
@@ -26,3 +26,27 @@ type Transport interface {
...
 }
+
+// TransportWithCreateOnly is a subtype of Transport that knows how to handle
+// the create-only option.
+type TransportWithCreateOnly interface {
+	Transport
+
+	// WrapObjectsHavingCreateOnly gets slice of Wrapee and wraps them into a single wrapped object.
+	// In case slice is empty, the function should return an empty wrapped object.
+	WrapObjectsHavingCreateOnly(objects []Wrapee) runtime.Object
```
Comment (vMaroon) [inline pkg/transport/transport.go:37]: Does this function filter out non create-only wrapees or it expects the list to be of create-only enabled wrapees? The behavior is not clear.

**Review by vMaroon — COMMENTED** [review]
Does the introduction of new transport type to solve the problem imply that some transports may implement the `Transport` interface and not the new one?

**Comment by MikeSpreitzer** [comment]
@vMaroon: I see that the transition plan was not obvious. I have pushed a revision that includes an explicit outline of the plan. That should address the points you raised above.

**Comment by MikeSpreitzer** [comment]
A little more detail on the problem: this repo (ks/ks) is tested together with the latest release of ks/OTP --- which does not implement the new wrapping method. So we have to break it into multiple steps, where each step does not break the other repo and building creates a system that works end-to-end (at least for the older functionality).

File: pkg/transport/generic_transport_controller.go (hunk 2 above)
Comment (vMaroon) [reply pkg/transport/generic_transport_controller.go:862]: Resolved

File: pkg/transport/generic_transport_controller.go (hunk 1 above)
Comment (vMaroon) [reply pkg/transport/generic_transport_controller.go]: Resolved

**Comment by vMaroon** [comment]
Thanks @MikeSpreitzer that makes more sense.
There is 1 concern in the "unresolved" comment then good to go from my side.

File: pkg/transport/transport.go (hunk 3 above)
Comment (MikeSpreitzer) [reply pkg/transport/transport.go:46]: The intent is that this method honors the create-only bit, correctly handling either value.

Comment (MikeSpreitzer) [reply pkg/transport/transport.go:49]: I pushed a revision that spells this out.

Comment (vMaroon) [reply pkg/transport/transport.go:49]: Thanks. So this means that once the transition is complete, the name will revert back to `WrapObjects`?

Comment (MikeSpreitzer) [reply pkg/transport/transport.go:49]: I was not planning on that but we can make that transition too with additional releases.

**Pull request merged**